🔐 Add JWT authentication with login endpoint

Volodymyr-Mykychak opened the pull request:
🔐 JWT Authentication Support

Implemented features:
- Added `POST /api/auth/login` endpoint returning JWT
- Created `AuthenticationService`, `JwtUtil`, and `JwtAuthenticationFilter`
- Integrated JWT into `SecurityFilterChain`
- Ensured `/auth/login` is publicly accessible
- Added request validation for login DTO
- Set JWT properties in both main and test property files

Volodymyr-Mykychak committed:
🔐 Add JWT authentication with login endpoint

liliia-ponomarenko commented on src/main/java/com/store/book/config/SecurityConfig.java:
```diff
@@ -30,14 +34,34 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
+                                               .requestMatchers(
+                                                       HttpMethod.POST,
+                                                       "/auth/registration",
+                                                       "/auth/login"
```
```suggestion
                                                       "/auth/**"
```

liliia-ponomarenko commented on src/main/java/com/store/book/security/AuthenticationService.java:
```diff
@@ -0,0 +1,27 @@
+                        request.email(),
+                        request.password()
+                )
+                                                                          );
```
```suggestion
                ));
```

liliia-ponomarenko commented on src/main/java/com/store/book/security/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,32 @@
+                                                                                    .getName()
+                                                                                    .name())
+                                                                            .toArray(String[]::new))
+                                                                 .build();
```
```suggestion
    public UserDetails loadUserByUsername(String username) {
        return userRepository.findByEmail(username).orElseThrow(
                () -> new EntityNotFoundException("There is no user with email: " + username));
```
Let's simplify it

Volodymyr-Mykychak replied:
Thank you for your suggestion 🙌

I kept the method `loadUserByUsername(String email)` with the mapping of roles into `UserDetails` because:
- Spring Security relies on `UserDetails` to retrieve authorities (`getAuthorities`) for access control (`hasRole(...)`, `hasAuthority(...)`)
- If I return the `User` entity directly using `findByEmail(...)` without mapping roles, it leads to issues with authorization (403 Forbidden)
- In that case, I’d need to implement the `GrantedAuthority` interface in the `Role` model and manually configure authority handling with `ROLE_...`

So applying the proposed change will require refactoring the `Role` and `User` models, as well as security configuration and tests.

Would it be acceptable to keep the current implementation? 🙂

liliia-ponomarenko commented on src/main/java/com/store/book/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearToken = request.getHeader("Authorization");
```
HttpHeader.AUTHORIZATION

liliia-ponomarenko commented on src/main/java/com/store/book/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,48 @@
+
+    private String getToken(HttpServletRequest request) {
+        String bearToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearToken) && bearToken.startsWith("Bearer ")) {
```
Make constant and use the method length

liliia-ponomarenko reviewed (changes requested):
Good job! A few comments ;)

Volodymyr-Mykychak committed:
🔧 Refactor: fix minor code style issues & improve security filter

- Simplified permitAll() to '/auth/**'
- Adjusted indentation for authenticate() call
- Extracted AUTHORIZATION header constant in JwtAuthenticationFilter
- Replaced magic number with Bearer prefix length method

Elena-Bruyako commented on src/main/java/com/store/book/config/SecurityConfig.java:
```diff
@@ -30,14 +34,33 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
+                                               .anyRequest().authenticated()
+                                      )
+                .addFilterBefore(
+                        jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter.class);
```
```suggestion
                .addFilterBefore(
                        jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter.class)
                        .userDetailsService(service)
```

Elena-Bruyako commented on src/main/java/com/store/book/security/AuthenticationService.java:
```diff
@@ -0,0 +1,26 @@
+                        request.email(),
+                        request.password()
+                ));
+        String token = jwtUtil.generateToken(request.email());
```
```suggestion
        String token = jwtUtil.generateToken(authentication.getName());
```

Elena-Bruyako commented on src/main/java/com/store/book/repository/UserRepository.java:
```diff
@@ -3,9 +3,14 @@
     Optional<User> findByEmail(String email);
 
+    @Query("SELECT u FROM User u JOIN FETCH u.roles WHERE u.email = :email")
+    Optional<User> findByEmailWithRoles(@Param("email") String email);
```
```suggestion
    @EntityGraph(attributePaths = "roles")
    Optional<User> findByEmail(String email);
```

Elena-Bruyako commented on src/main/java/com/store/book/security/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        User user = userRepository.findByEmailWithRoles(email)
```
see comment in userRepo
use findByEmail
set particular role during registration
User should implement UserDetails interface
Role should implement GrantedAuthority

remove this redundant logic and use Spring Security features

Volodymyr-Mykychak replied:
I’ve tried updating this service to return the domain User entity directly as discussed, but doing so breaks JWT-based authorization — Spring Security doesn’t correctly extract roles from the custom User object, leading to "Access Denied" errors for secured endpoints. Returning a built-in UserDetails object with explicit authorities (via .roles(...)) works reliably and ensures proper role mapping. I’ll revisit the suggested change if I find a way to preserve JWT compatibility with the domain User object.

Elena-Bruyako reviewed (changes requested):
see comments

Volodymyr-Mykychak committed:
🔐 Fix JWT role extraction issues by using built-in UserDetails in CustomUserDetailsService

- Reverted to using org.springframework.security.core.userdetails.User in CustomUserDetailsService
  due to JWT not recognizing roles from the domain User object.
- Ensures proper authority mapping and avoids Access Denied errors for secured endpoints.
- Added explanatory comment for reviewer regarding current limitation and workaround.

liliia-ponomarenko reviewed (approved):
Good job!
